File: DriverSource/Windows10IoTCore/TurtaRelayHAT/TurtaRelayHAT/RelayController.cs
/* Turta® Relay HAT Helper for Windows® 10 IoT Core
 * Copyright © 2017 Turta
 * Distributed under the terms of the MIT license.
 */

using System;
using Windows.Devices.Gpio;

namespace TurtaRelayHAT
{
    public class RelayController : IDisposable
    {
        #region Globals

        // GPIO Device
        private static GpioPin relay1, relay2, relay3, relay4;

        #endregion

        #region Constructor

        /// <summary>
        /// Initiates the relays to turn on / off devices up to DC24V / AC240V 5A each.
        /// </summary>
        public RelayController()
        {
            // Initiate the GPIO Controller.
            GpioController gpioController = GpioController.GetDefault();

            // Configure the pins.
            relay1 = gpioController.OpenPin(21);
            relay2 = gpioController.OpenPin(22);
            relay3 = gpioController.OpenPin(23);
            relay4 = gpioController.OpenPin(24);

            relay1.Write(GpioPinValue.Low);
            relay2.Write(GpioPinValue.Low);
            relay3.Write(GpioPinValue.Low);
            relay4.Write(GpioPinValue.Low);

            relay1.SetDriveMode(GpioPinDriveMode.Output);
            relay2.SetDriveMode(GpioPinDriveMode.Output);
            relay3.SetDriveMode(GpioPinDriveMode.Output);
            relay4.SetDriveMode(GpioPinDriveMode.Output);
        }

        #endregion

        #region Relay Control

        /// <summary>
        /// Controls the relay state.
        /// </summary>
        /// <param name="ch">Relay channel. 1 to 4.</param>
        /// <param name="state">Relay state. True for enable, false for disable.</param>
        public void SetRelay(int ch, bool state)
        {
            switch (ch)
            {
                case 1:
                    relay1.Write(state ? GpioPinValue.High : GpioPinValue.Low);
                    break;

                case 2:
                    relay2.Write(state ? GpioPinValue.High : GpioPinValue.Low);
                    break;

                case 3:
                    relay3.Write(state ? GpioPinValue.High : GpioPinValue.Low);
                    break;

                case 4:
                    relay4.Write(state ? GpioPinValue.High : GpioPinValue.Low);
                    break;

                default:
                    break;
            }
        }

        #endregion

        #region Disposal

        /// <summary>
        /// Disables the relays and then cleans up the resources.
        /// </summary>
        public void Dispose()
        {
            relay1.Write(GpioPinValue.Low);
            relay2.Write(GpioPinValue.Low);
            relay3.Write(GpioPinValue.Low);
            relay4.Write(GpioPinValue.Low);

            relay1.Dispose();
            relay2.Dispose();
            relay3.Dispose();
            relay4.Dispose();
        }

        #endregion
    }
}

File: DriverSource/Raspbian/Python/RelayController.py
# Turta Relay HAT Helper for Raspbian
# Distributed under the terms of the MIT license.

import RPi.GPIO as GPIO

#Pins
relay1, relay2, relay3, relay4 = 21, 22, 23, 24

#Initialize
def Init():
    GPIO.setwarnings(False)
    GPIO.setmode(GPIO.BCM)
    GPIO.setup(relay1, GPIO.OUT)
    GPIO.setup(relay2, GPIO.OUT)
    GPIO.setup(relay3, GPIO.OUT)
    GPIO.setup(relay4, GPIO.OUT)
    GPIO.output(relay1, GPIO.LOW)
    GPIO.output(relay2, GPIO.LOW)
    GPIO.output(relay3, GPIO.LOW)
    GPIO.output(relay4, GPIO.LOW)
    return

#Relay Control
def SetRelay(ch, st):
    if (ch == 1):
        GPIO.output(relay1, GPIO.HIGH if st else GPIO.LOW)
    elif (ch == 2):
        GPIO.output(relay2, GPIO.HIGH if st else GPIO.LOW)
    elif (ch == 3):
        GPIO.output(relay3, GPIO.HIGH if st else GPIO.LOW)
    elif (ch == 4):
        GPIO.output(relay4, GPIO.HIGH if st else GPIO.LOW)
    return

#Relay Readout
def ReadRelayState(ch):
    if (ch == 1):
        return GPIO.input(relay1)
    elif (ch == 2):
        return GPIO.input(relay2)
    elif (ch == 3):
        return GPIO.input(relay3)
    elif (ch == 4):
        return GPIO.input(relay4)

#Release The Resources
def Dispose():
    GPIO.cleanup()
    return
